### PR TITLE
feat(ssa): Switch mem2reg pass to be per function rather than per block

### LINKED ISSUE #2218: SSA: Enable mem2reg to be per function rather than per block
### Problem

The mem2reg pass is currently done in a single block context. This means that values which contain loads whose results are known from store instructions in previous blocks will not be removed. This causes issues in a couple situations:

In https://github.com/noir-lang/noir/issues/1889, merging of slices occurs during flattening of the CFG. However, mem2reg currently occurs after flattening (because we can perform mem2reg on one flattened block). Thus, we cannot merge slices in flattening as we have not loaded the array value that was stored in a previous block. Being able to resolve loads from stores in previous blocks will enable us to handle merge slices in flattening.

In https://github.com/noir-lang/noir/issues/1035 as per https://github.com/noir-lang/noir/issues/1035#issuecomment-1650132409 moving the mem2reg pass before unrolling fixes the issue, however, due to mem2reg being per block rather than per function other bugs are not caught. Firstly, the issue of slices not being correctly loaded for merge_slices still exists. Moving the mem2reg pass before unrolling also causes this snippet to be incorrect:
```
fn main(x : Field, y : pub Field) {
    let q = &mut &mut 0;
    let q1 = *q;
    let q2 = *q;

    if x != y {
        *q1 = 1;
        // mem2reg will know q1 = 1, but not that q2 = 1
        *q2 = 2; // now we'd expect q1 == q2 == 2, but mem2reg now thinks q1 = 1 & q2 = 2
        
        // This now passes and we've broken mem2reg :(
        assert(*q1 == 1);
    }
}
```
We should expect `assert(*q1 == 1)` to fail. This failure occurs with how the SSA is currently set-up but with the limitations listed above.

### Happy Case

We should be able to easily reorder mem2reg in the SSA. This will require mem2reg to run with a context per function rather than per block. After moving `mem2reg` before unrolling, the current functionality of the SSA should not be reduced and the issues listed in the **Problem** section should be addressed. 

### Alternatives Considered

N/A

In order for merge_slices to still occur in flattening there is really no other option than being able to resolve loads from stores in previous blocks. Otherwise, flattening of slice would start to get unnecessarily complex. Switching to per function for mem2reg will potentially also fix other bugs along with providing functionality for the merger of slices. 

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

Update mem2reg to be ran per function rather than per block. This enables mem2reg to be easily reordered amongst the SSA pass, such as being placed before loop unrolling. The previous per block mem2reg would break if placed before the flattening of the CFG.

## Problem\*

Resolves #2218

## Summary\*

The main change in this PR is that we now have to analyze the CFG when we want to find a load to substitute and fetch the respective load value to be substituted. Three new methods have been made:

`fetch_load_value_to_substitute` 
-  This method fetches the already save load values to substitute for a given address
    The search starts at the block supplied as a parameter. If there is not a load to substitute
    the CFG is analyzed to determine whether a predecessor block has a load value to substitute.
    If there is no load value to substitute the original address is returned.

`find_load_to_substitute`
- This method determines which loads should be substituted.
    Starting at the block supplied as a parameter, we check whether a store has occurred with the given address.
    If no store has occurred in the supplied block, the CFG is analyzed to determine whether
    a predecessor block has a store at the given address.

`has_common_successor`
- This method checks whether the block we are examining shares a common successor with its predecessor block. This check is necessary to prevent removing stores across blocks whose execution is dependent upon a runtime if statement. Without this check we could potentially remove a store and replace its load for only one block of a jmp if. We should be waiting for flattening to handle both scenarios. Thus, the mem2reg pass should also be ran after flattening to handle any runtime if statements. 

We maintain two maps for loads to substitute. Per function, and per block. As when it comes time to actually substitute load result values and delete load instructions we want it to be done in the correct block order. We could most likely come-up with a strategy to use the per function loads to substitute map for replacing load result values, but simply maintaining a separate map felt like a simpler approach. 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs [hunk 1]
```diff
@@ -122,21 +146,115 @@ impl PerBlockContext {
...
+                load_values_to_substitute_per_block.insert(result_value, *last_value);
+                self.load_values_to_substitute_per_func.insert(result_value, (*last_value, block));
+                return true;
+            }
+
+            // If no stores have been found in the current block, check the block's predecessors.
+            let predecessors = self.cfg.predecessors(block);
+            for predecessor in predecessors {
+                // TODO: Do I need is_reachable here? We are looping over only the reachable blocks but does
+                // that include a reachable block's predecessors?
```
Comment (vezenovm) [inline crates/noirc_evaluator/src/ssa/opt/mem2reg.rs:246]: Wasn't sure if I needed is_reachable here

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs [hunk 2]
```diff
@@ -122,21 +146,115 @@ impl PerBlockContext {
...
+        address: ValueId,
+        dfg: &DataFlowGraph,
+        instruction_id: &InstructionId,
+        loads_to_substitute: &mut BTreeMap<InstructionId, ValueId>,
+        load_values_to_substitute_per_block: &mut BTreeMap<ValueId, ValueId>,
+    ) -> bool {
+        let mut stack = vec![block_id];
+        let mut visited = HashSet::new();
+
+        let mut dom_tree = DominatorTree::with_cfg_and_post_order(&self.cfg, &self.post_order);
```
Comment (vezenovm) [inline crates/noirc_evaluator/src/ssa/opt/mem2reg.rs:216]: I will submit an update in a moment to move this out as the dom tree is being recomputed per call to find_load_substitute which is inefficient.

Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/opt/mem2reg.rs]: Fixed now

File: crates/nargo_cli/tests/execution_success/references/src/main.nr [hunk 3]
```diff
@@ -1,4 +1,6 @@
 fn main(mut x: Field) {
+    regression_2218(x, 10);
```
Comment (jfecher) [inline crates/nargo_cli/tests/execution_success/references/src/main.nr:2]: Can you move this to the end with the other regressions?

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs [hunk 4]
```diff
@@ -1,101 +1,129 @@
...
+use super::unrolling::{find_all_loops, Loops};
+
 impl Ssa {
     /// Attempts to remove any load instructions that recover values that are already available in
-    /// scope, and attempts to remove store that are subsequently redundant, as long as they are
-    /// not stores on memory that will be passed into a function call or returned.
+    /// scope, and attempts to remove stores that are subsequently redundant.
+    /// As long as they are not stores on memory used inside of loops
     pub(crate) fn mem2reg(mut self) -> Ssa {
+        dbg!("start mem2reg");
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/mem2reg.rs:27]:
```suggestion
```

File: crates/noirc_evaluator/src/ssa/opt/unrolling.rs [hunk 5]
```diff
@@ -59,20 +59,20 @@ struct Loop {
     blocks: HashSet<BasicBlockId>,
 }
 
-struct Loops {
+pub(crate) struct Loops {
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/unrolling.rs:62]: Several items here were changed to `pub(crate)`, can it be reverted?

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs [hunk 6]
```diff
@@ -122,21 +150,117 @@ impl PerBlockContext {
...
+
+        while let Some(block) = stack.pop() {
+            visited.insert(block);
+
+            for l in self.loops.yet_to_unroll.iter() {
+                // We do not want to substitute loads that take place within loops as this pass
+                // can occur before loop unrolling
+                if block == l.header {
+                    return false;
+                }
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/mem2reg.rs:231]: If this pass can't handle loads within loops, will it also break if a load is used after a loop which mutates it?
```rs
fn main() {
    let ref = &mut 0;
    for _ in 0 .. 10 {
        *ref += 1;
    }
    assert(*ref == 10);
}
```
Or variations of the above?

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs (hunk 1 above)
Comment (jfecher) [reply crates/noirc_evaluator/src/ssa/opt/mem2reg.rs:252]: It could theoretically be necessary if there were previous blocks that were removed but still jump to any of these blocks internally. I'm not sure if this ever occurs on any of our optimization passes but it's safer to have the check I suppose.

Can `PostOrder` be used at all since we're iterating backwards? Or not because we have to start on a given block id?

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs [hunk 7]
```diff
@@ -1,101 +1,129 @@
...
             // function's blocks, it is safe to remove any stores that do not touch such
             // allocations.
-            for context in contexts {
-                context.remove_unused_stores(&mut function.dfg, &all_protected_allocations);
+            for block in function.reachable_blocks() {
+                context.remove_unused_stores(&mut function.dfg, &all_protected_allocations, block);
             }
         }
-
+        dbg!("ended mem2reg");
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/opt/mem2reg.rs:57]:
```suggestion
```

**Review by jfecher — CHANGES_REQUESTED** [review]
I haven't finished a full review yet - I'll have to look at this more in depth tomorrow. I'd like to test it on some examples as well.

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs (hunk 6 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/opt/mem2reg.rs]: It would break if we also removed the mem2reg pass that occurs after flattening. This snippet passes as we run mem2reg before loop unrolling and also after flattening. I have actually updated this as I found bug while working on merge slices which is based off of this PR. I no longer construct a `Loops` object and instead only continue searching for stores on blocks with one predecessor. This enables the mem2reg to actually be fully re-orderable because if we remove loads/stores in blocks with multiple predecessors we risk overwriting other comptime values that have been simplified in other passes such as loop unrolling or flattening. 

For example this snippet:
```
    let mut new_slice = [];
    for i in 0..5 {
        new_slice = new_slice.push_back(i);
    }
    assert(new_slice.len() == 5);
```
will provide this SSA after inlining:
```
After Inlining:
acir fn main f1 {
  b0():
    v0 = allocate
    store [] at v0
    jmp b1(Field 0)
  b1(v2: Field):
    v5 = lt v2, Field 5
    jmpif v5 then: b2, else: b3
  b2():
    v10 = load v0
    v12 = call slice_push_back(v10, v2)
    store v12 at v0
    v14 = add v2, Field 1
    jmp b1(v14)
  b3():
    v6 = load v0
    v8 = call array_len(v6)
    v9 = eq v8, Field 5
    constrain v9
    return 
}
```
If we were load v0 in block 2, the first argument to `slice_push_back` would be replaced with the constant `[]`. 
Bad SSA:
```
acir fn main f1 {
  b0():
    v0 = allocate
    jmp b1(Field 0)
  b1(v2: Field):
    v5 = lt v2, Field 5
    jmpif v5 then: b2, else: b3
  b2():
    v12 = call slice_push_back([], v2)
    v14 = add v2, Field 1
    jmp b1(v14)
  b3():
    v8 = call array_len([])
    v9 = eq v8, Field 5
    constrain v9
    return 
}
```
While in fact we should only remove the load once loop unrolling is done and there are five `slice_push_back` calls, so that only the first value is replaced.

File: crates/noirc_evaluator/src/ssa/opt/mem2reg.rs (hunk 1 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/opt/mem2reg.rs]: I am kind of using `PostOrder` by using the dom tree no?

File: crates/nargo_cli/tests/execution_success/references/src/main.nr (hunk 3 above)
Comment (vezenovm) [reply crates/nargo_cli/tests/execution_success/references/src/main.nr]: done

File: crates/nargo_cli/tests/execution_success/references/src/main.nr [hunk 8]
```diff
@@ -35,6 +35,12 @@ fn main(mut x: Field) {
     regression_2054();
     regression_2030();
     regression_2255();
+
+    // x == 3 at this point
+    regression_2218_if_inner_if(x, 10);
+    regression_2218_if_inner_else(20, x);
+    regression_2218_else(x, 3);
+    regression_2218_loop(x, 10);
```
Comment (jfecher) [inline crates/nargo_cli/tests/execution_success/references/src/main.nr:43]: Shouldn't these arguments be inside `if_inner_if`, etc if we want to guarantee it takes the inner_if path?

File: crates/nargo_cli/tests/execution_success/regression/src/main.nr [hunk 9]
```diff
@@ -76,6 +76,7 @@ fn main(x: [u8; 5], z: Field)
     //Issue 1144
 	let (nib, len) = compact_decode(x,z);
 	assert(len == 5);
+	// dep::std::println(nib);
```
Comment (jfecher) [inline crates/nargo_cli/tests/execution_success/regression/src/main.nr:79]:
```suggestion
```

**Review by jfecher — COMMENTED** [review]
I'm still not 100% confidant in this pass honestly since it is getting harder to keep the corner cases in my head as the pass gets more complex but my thoughts are we should merge it (after some minor comments) since it passes all the tests and flag any issues for fixing later. I have to make some changes to mem2reg to support alias analysis now as well so I'll be making some other large changes to the pass in the future.

File: crates/nargo_cli/tests/execution_success/references/src/main.nr (hunk 8 above)
Comment (vezenovm) [reply crates/nargo_cli/tests/execution_success/references/src/main.nr:43]: Sorry not sure what you mean here do you mean the loop test inside of the others?

Comment (jfecher) [reply crates/nargo_cli/tests/execution_success/references/src/main.nr:43]: I mean `regression_2218_if_inner_if();` Then remove the `x` and `y` parameters inside that function and replace them with `x` and `10` to guarantee that function takes the correct path in `regreesion_2218(..)`

Comment (jfecher) [reply crates/nargo_cli/tests/execution_success/references/src/main.nr:43]: Since as it is right now you can do `regression_2218_if_inner_if(20, 20)` and not get the if_inner_if path.

Comment (jfecher) [reply crates/nargo_cli/tests/execution_success/references/src/main.nr:43]: Ah, fair enough. I missed the witness part.

**Review by jfecher — APPROVED** [review]
👍

File: crates/nargo_cli/tests/execution_success/references/src/main.nr (hunk 8 above)
Comment (vezenovm) [reply crates/nargo_cli/tests/execution_success/references/src/main.nr:43]: I included an assert to guarantee we hit the if statements we expect

**Comment by vezenovm** [comment]
@jfecher Could you approve once more?

**Review by jfecher — APPROVED** [review]

**Review by jfecher — APPROVED** [review]
